compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs

```diff
@@ -871,15 +895,24 @@ impl<'block> BrilligBlock<'block> {
                 self.brillig_context.mov_instruction(target_size, source_size_as_register);
                 self.brillig_context.usize_const(target_rc, 1_usize.into());
             }
-            _ => unreachable!("ICE: array set on non-array"),
+            _ => unreachable!("ICE: {method_str} on non-array"),
         }
 
-        // Then set the value in the newly created array
-        self.store_variable_in_array(
-            destination_pointer,
-            SingleAddrVariable::new_usize(index_register),
-            value_variable,
-        );
+        match index_and_value_or_len {
+            Ok((index_register, value_variable)) => {
+                // Then set the value in the newly created array
+                self.store_variable_in_array(
+                    destination_pointer,
+                    SingleAddrVariable::new_usize(index_register),
+                    value_variable,
+                );
+            }
+            Err(destination_len_variable) => {
+                // we need to explicitly set the destination_len_variable
+                self.brillig_context
```

Comment:
Could we do this `mov` inside the `as_slice` intrinsic? If we want to combine usage of `convert_ssa_array_set` we can then do an option for `(index_register, value_variable)` about whether we are actually setting a value in the newly created array